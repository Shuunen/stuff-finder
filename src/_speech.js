const App = require('./app')

class AppSpeech extends App {
  constructor () {
    super('speech')
    this.log('constructor')
    this.initRecognition()
    this.intent = ''
  }

  initRecognition () {
    const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition
    const recognition = new SpeechRecognition()
    recognition.lang = 'fr-FR'
    recognition.interimResults = false
    recognition.maxAlternatives = 1
    recognition.onresult = (event) => {
      const result = event.results[event.results.length - 1][0]
      this.onRecognition(result.transcript, result.confidence)
    }
    recognition.onspeechend = () => (console.log('speech end') && recognition.stop())
    recognition.onnomatch = (event) => console.log('onnomatch')
    recognition.onerror = (event) => console.log('error occurred in recognition: ' + event.error)
    this.recognition = recognition
  }

  onRecognition (sentence, confidence) {
    this.log('sentence :', sentence)
    this.log('confidence : ' + confidence)
    this.emit('do-prompt-intent', { stuff: sentence, intent: this.intent })
  }

  start (intent = 'unknown') {
    this.intent = intent
    this.recognition.start()
    console.log('listening...')
  }
}

window.appSpeech = new AppSpeech()
