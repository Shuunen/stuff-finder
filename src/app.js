class App {
  constructor (name = 'unknown') {
    this.name = name
    window.onload = () => this.beforeLoad()
  }

  log () {
    console.log.apply(console.log, [this.name, ':', ...arguments])
  }

  emit (eventName, eventData) {
    console.log(`%c${eventName}`, 'color: blue', eventData)
    window.dispatchEvent(new CustomEvent(eventName, { detail: eventData }))
  }

  on (eventName, callback) {
    window.addEventListener(eventName, event => callback.bind(this)(event.detail))
  }

  beforeLoad () {
    this.setupElements()
    this.setupListeners()
    this.onLoad()
  }

  onLoad () { }

  setupElements () { }

  setupListeners () { }
}

module.exports = App
