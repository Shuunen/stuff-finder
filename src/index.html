<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Stuff Finder</title>
  <meta name="description" content="Sorting things is pointless if you can't find them afterwards :D">
  <meta name="theme-color" content="#209CEE" />
  <link rel="apple-touch-icon" sizes="180x180" href="assets/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="assets/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="assets/favicon-16x16.png">
  <link rel="stylesheet" type="text/css" media="screen" href="styles/main.css">
  <link rel="manifest" href="assets/site.webmanifest">
</head>

<body class="stuff-finder col center middle">

  <app-toaster></app-toaster>
  <app-loader></app-loader>
  <app-prompter></app-prompter>
  <app-scan-code></app-scan-code>
  <app-search-button></app-search-button>
  <app-search-results></app-search-results>
  <app-print-barcodes></app-print-barcodes>
  <app-settings-trigger></app-settings-trigger>

  <app-modal name="settings">
    <app-form name="settings" title="Settings">
      <label>
        <a href="//airtable.com/api" target="_blank" rel="noreferrer">Airtable api base</a>
        <input name="base" placeholder="your api base" pattern="^app\w{14}$" required maxlength="17" autofocus>
      </label>
      <label>
        <a href="//airtable.com/account" target="_blank" rel="noreferrer">Airtable api key</a>
        <input name="key" placeholder="your api key" pattern="^key\w{14}$" required maxlength="17">
      </label>
      <label>
        Airtable table name
        <input name="table" pattern="\S+" placeholder="your table name" required>
      </label>
      <label>
        Airtable view name
        <input name="view" pattern="\S+" placeholder="your table name" required value="stuff-finder">
      </label>
    </app-form>

    <details class="mb1 left">
      <summary>What this app do with my credentials ?</summary>
      <small>
        <ul>
          <li>fetch your Airtable data</li>
          <li>store credentials locally in your browser</li>
          <li>nothing else</li>
        </ul>
        Feel free to check this app source code on the <a href="//github.com/Shuunen/stuff-finder" target="_blank" rel="noreferrer">Github project
          page</a>.
      </small>
    </details>
  </app-modal>

  <script src="main.js"></script>
</body>

</html>
