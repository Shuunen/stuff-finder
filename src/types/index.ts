/* eslint-disable sonar/no-wildcard-import */
export * from './events'
export * from './item'
export * from './print'
export * from './requests'
export * from './settings'

