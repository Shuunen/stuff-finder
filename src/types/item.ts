export interface ItemThumbnail {
  url: string
  width: number
  height: number
}

export interface ItemPhoto {
  filename: string
  height: number
  id: string
  size: number
  type: string
  url: string
  width: number
  thumbnails: {
    small: ItemThumbnail
    large: ItemThumbnail
    full: ItemThumbnail
  }
}

export const enum ItemStatus {
  acheté = 'acheté',
  défectueux = 'défectueux',
  donné = 'donné',
  jeté = 'jeté',
  renvoyé = 'renvoyé',
  vendu = 'vendu',
}

export const enum ItemField {
  barcode = 'barcode',
  box = 'box',
  brand = 'brand',
  category = 'category',
  details = 'details',
  drawer = 'drawer',
  id = 'id',
  location = 'location',
  name = 'name',
  photo = 'photo',
  price = 'price',
  reference = 'reference',
  referencePrinted = 'ref-printed',
  status = 'status',
  updatedOn = 'updated-on',
}

export interface Item {
  [ItemField.barcode]: string
  [ItemField.box]: string
  [ItemField.brand]: string
  [ItemField.category]: string
  [ItemField.details]: string
  [ItemField.drawer]: string
  [ItemField.id]: string
  [ItemField.location]: string
  [ItemField.name]: string
  [ItemField.photo]?: ItemPhoto[]
  [ItemField.price]?: number
  [ItemField.reference]: string
  [ItemField.referencePrinted]: boolean
  [ItemField.status]: ItemStatus
  [ItemField.updatedOn]: string
}

export type ItemSuggestions = Record<keyof Item, string[]>

export interface CommonLists {
  boxes: string[]
  locations: string[]
  statuses: string[]
  drawers: string[]
  categories: string[]
}
