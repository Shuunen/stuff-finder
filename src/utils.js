export const dom = (type, content = '', classes = '') => {
  const element = document.createElement(type)
  element.className = classes
  element.innerHTML = content
  return element
}

export const div = (classes, content = '') => dom('div', content, classes)

export const img = (alt, source, classes = '') => {
  const element = dom('img', '', classes)
  element.alt = alt
  element.src = source
  return element
}

export const button = (content, classes = '') => {
  return dom('button', content, `bg-blue-800 m-auto sm:ml-0 px-4 py-1 ${classes}`)
}

export const p = (content, classes = '') => dom('p', content, classes)
