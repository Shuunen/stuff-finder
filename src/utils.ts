import { dom, emit } from 'shuutils'

export const button = (content: string, classes = '', secondary = false): HTMLButtonElement => {
  const theme = secondary ? 'hover:opacity-100 opacity-80' : 'from-purple-500 to-purple-700 text-white hover:from-purple-700 hover:to-purple-900'
  const button_ = dom('button', `text-lg h-10 max-w-xs px-12 md:text-base bg-gradient-to-tr shadow-md hover:shadow-lg rounded transition-all duration-200 each-in-out sm:px-6 ${theme} ${classes}`, content)
  button_.type = 'button'
  return button_
}

export const showError = (message: string) => {
  console.error(message)
  emit('app-toaster--show', { type: 'error', message })
}

export const showLog = (message: string, data = '') => {
  console.log(message, data)
  emit('app-toaster--show', { type: 'info', message })
}
