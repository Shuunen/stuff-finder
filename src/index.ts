import Fuse from 'fuse.js'
import { copy, emit, fillTemplate, on, pickOne, sanitize, storage } from 'shuutils'
import './assets/styles.min.css'
import './components'
import { EMPTY_APP_SETTINGS, EMPTY_COMMON_LISTS, EMPTY_ITEM } from './constants'
import './services'
import type { AirtableRecord, AirtableResponse, AppActionEvent, AppClearCacheEvent, AppClearCredentialsEvent, AppFormEditItemSaveEvent, AppFormSettingsErrorEvent, AppFormSettingsReadyEvent, AppFormSettingsSaveEvent, AppFormSettingsSetEvent, AppLoaderToggleEvent, AppModalAddItemCloseEvent, AppModalEditItemCloseEvent, AppModalSearchResultsCloseEvent, AppModalSettingsCloseEvent, AppPrompterTypeEvent, AppScanCodeStartEvent, AppSettings, AppSettingsTriggerAnimateEvent, AppSpeechStartEvent, AppStatusEvent, AppToasterShowEvent, CommonLists, FormEditFormData, Item, ItemPhoto, ItemsReadyEvent, SearchOrigin, SearchResultsEvent, SearchRetryEvent, SearchStartEvent } from './types'
import { ItemField, ItemStatus } from './types'
import { find, logger, patch, post, valuesToOptions } from './utils'

class App {

  private apiUrl = ''

  private lastSearchOrigin: SearchOrigin = 'default'

  private items: Item[] = []

  private fuse!: Fuse<Item>

  private commonListsLoaded = false

  public constructor () {
    logger.log('app start')
    storage.prefix = '@shuunen/stuff-finder_'
    on<AppFormSettingsSaveEvent>('app-form--settings--save', this.onSettingsSave.bind(this))
    on<AppFormEditItemSaveEvent>('app-form--edit-item--save', this.onEditItem.bind(this))
    on<SearchStartEvent>('search-start', this.onSearchStart.bind(this))
    on<SearchRetryEvent>('search-retry', this.onSearchRetry.bind(this))
    on<AppClearCacheEvent>('app-clear-cache', this.onClearCache.bind(this))
    on<AppClearCredentialsEvent>('app-clear-credentials', this.onClearCredentials.bind(this))
    this.checkExistingSettings()
    this.showTitle()
    this.handleActions()
  }

  private onClearCredentials (): void {
    logger.log('clearing credentials...')
    storage.clear('app-settings')
  }

  private onClearCache (): void {
    logger.log('clearing cached items...')
    storage.clear('items')
    document.location.reload()
  }

  private checkExistingSettings (): void {
    const settings = storage.get<AppSettings>('app-settings', EMPTY_APP_SETTINGS)
    if (!settings.base) { this.settingsActionRequired(true); return }
    void this.onSettingsSave(settings)
    on<AppFormSettingsReadyEvent>('app-form--settings--ready', () => emit<AppFormSettingsSetEvent>('app-form--settings--set', settings))
  }

  private coolAscii (): string {
    return pickOne(['( ＾◡＾)', '♥‿♥', '八(＾□＾*)', '(◡ ‿ ◡ ✿)', '(=^ェ^=)', 'ʕ •ᴥ•ʔ', '(*°∀°)', '\\(-ㅂ-)/', 'ლ(╹◡╹ლ)', 'ლ(o◡oლ)', '＼(＾O＾)／'])
  }

  private showTitle (): void {
    emit<AppPrompterTypeEvent>('app-prompter--type', ['Stuff Finder', 300, `Stuff Finder\n${this.coolAscii()}`])
  }

  private async onSettingsSave (settings: AppSettings): Promise<void> {
    this.apiUrl = `https://api.airtable.com/v0/${settings.base}/${settings.table}?api_key=${settings.key}&view=${settings.view}`
    const itemsLoaded = await this.loadItems()
    if (!itemsLoaded) {
      this.settingsActionRequired(true, 'failed to use api settings')
      return
    }
    this.settingsActionRequired(false)
    emit<AppModalSettingsCloseEvent>('app-modal--settings--close')
    if (this.items.length > 0) emit<ItemsReadyEvent>('items-ready')
    storage.set('app-settings', settings)
  }

  private settingsActionRequired (actionRequired: boolean, errorMessage = ''): void {
    emit<AppSettingsTriggerAnimateEvent>('app-settings-trigger--animate', actionRequired)
    emit<AppFormSettingsErrorEvent>('app-form--settings--error', errorMessage)
    emit<AppStatusEvent>('app-status', actionRequired ? 'settings-required' : 'ready')
    this.isLoading(false)
  }

  private isLoading (active: boolean): void {
    logger.log('isLoading active ?', active)
    emit<AppLoaderToggleEvent>('app-loader--toggle', active)
  }

  private async loadItems (): Promise<boolean> {
    this.isLoading(true)
    const cachedItems = storage.get<Item[]>('items', [])
    if (cachedItems.length === 0) logger.showLog('no cached items found')
    let response = await this.fetchApi()
    if (response.error) {
      this.isLoading(false)
      logger.showError(`airtable fetch failed with error : ${response.error.message}`)
      return false
    }
    let records = response.records
    if (!records[0] || records.length === 0) {
      this.isLoading(false)
      logger.showError('airtable fetch returned no records')
      return false
    }
    const remote = this.airtableRecordToItem(records[0])
    if (cachedItems.some(item => item.id === remote.id && item[ItemField.updatedOn] === remote[ItemField.updatedOn])) {
      this.items = cachedItems
      logger.showLog('no updates from Airtable, cache seems up to date')
      this.initFuse()
      return true
    }
    let offset = response.offset
    while ((offset?.length ?? 0) > 0) {
      response = await this.fetchApi(offset)
      offset = response.offset
      records = [...records, ...response.records]
    }
    this.parseApiRecords(records)
    return true
  }

  private async fetchApi (offset?: string): Promise<AirtableResponse> {
    const sortByUpdatedFirst = '&sort%5B0%5D%5Bfield%5D=updated-on&sort%5B0%5D%5Bdirection%5D=desc'
    const url = this.apiUrl + (offset !== undefined ? `&offset=${offset}` : '') + sortByUpdatedFirst
    if (!url.startsWith('https://api.airtable.com/v0/')) throw new Error('invalid api url')
    return fetch(url).then(async response => response.json()) as Promise<AirtableResponse>
  }

  private parseApiRecords (records: AirtableRecord[]): void {
    // this.showLog('parsing api records :', records )
    const boxes: string[] = []
    const locations: string[] = []
    const statuses = [ItemStatus.acheté, ItemStatus.vendu, ItemStatus.donné, ItemStatus.renvoyé, ItemStatus.défectueux, ItemStatus.jeté]
    const categories: string[] = []
    const drawers = ['', '1', '2', '3', '4', '5', '6', '7']
    records.forEach(record => {
      const location = record.fields.location ?? ''
      if (location.length > 0 && !locations.includes(location)) locations.push(location)
      const box = record.fields.box ?? ''
      if (box.length > 0 && !boxes.includes(box)) boxes.push(box)
      const status = record.fields.status ?? ItemStatus.acheté
      if (!statuses.includes(status)) statuses.push(status)
      const category = record.fields.category ?? ''
      if (category.length > 0 && !categories.includes(category)) categories.push(category)
    })
    this.saveCommonLists({ boxes, locations, statuses, categories, drawers })
    this.items = records.map(record => this.airtableRecordToItem(record))
    logger.showLog(`${this.items.length} item(s) loaded ` + this.coolAscii())
    this.initFuse()
  }

  private initFuse (): void {
    if (!this.readCommonLists()) return
    // https://fusejs.io/
    const options: Fuse.IFuseOptions<Item> = {
      distance: 200, // see the tip at https://fusejs.io/concepts/scoring-theory.html#scoring-theory
      threshold: 0.35, // 0 is perfect match
      ignoreLocation: true,
      getFn: (object: Item, path: string[] | string) => {
        const value = Fuse.config.getFn(object, path)
        if (Array.isArray(value)) return value.map((element: string) => sanitize(element))
        if (typeof value === 'string') return sanitize(value)
        return value
      },
      keys: [{
        name: ItemField.name,
        weight: 4,
      }, {
        name: ItemField.brand,
        weight: 2,
      }, {
        name: ItemField.details,
        weight: 4,
      }, {
        name: ItemField.category,
        weight: 1,
      }], // this is not generic ^^"
    }
    this.fuse = new Fuse(this.items, options)
    storage.set('items', this.items)
  }

  private onSearchStart (event: SearchStartEvent): void {
    const input = event.str.trim()
    this.lastSearchOrigin = event.origin
    const result = this.items.find(item => item.reference === input || item.barcode === input)
    const results = result ? [result] : this.fuse.search(sanitize(input)).map(item => item.item)
    let title = 'No results found'
    if (results.length > 0) title = results.length === 1 ? 'One result found' : `${results.length} results found`
    title += ` for “${input}”`
    const data: SearchResultsEvent = { title, results, byReference: Boolean(result), input, scrollTop: Boolean(event.scrollTop) }
    emit<SearchResultsEvent>('search-results', data)
  }

  private onSearchRetry (): boolean {
    logger.log('retry and this.lastSearchOrigin', this.lastSearchOrigin)
    if (this.lastSearchOrigin === 'type') {
      find.one<HTMLInputElement>('#input-type').focus()
      return true
    }
    if (this.lastSearchOrigin === 'scan') return emit<AppScanCodeStartEvent>('app-scan-code--start')
    if (this.lastSearchOrigin === 'speech') return emit<AppSpeechStartEvent>('app-speech--start')
    logger.showError('un-handled search retry case')
    return false
  }

  private airtableRecordToItem (record: AirtableRecord): Item {
    return {
      ...EMPTY_ITEM,
      ...record.fields,
      id: record.id,
    }
  }

  private async onEditItem (data: FormEditFormData): Promise<boolean> {
    const fields = this.getItemFieldsToPush(data)
    if (Object.keys(fields).length === 0) { logger.showLog('no changes to push'); return false }
    this.isLoading(true)
    const response = await this.pushItemRemotely(fields, data.id)
    logger.log('onEditItem response', response)
    this.isLoading(false)
    if (response.error) { logger.showError(response.error.message); return false }
    const item = this.airtableRecordToItem(response)
    this.pushItemLocally(item)
    emit<AppModalEditItemCloseEvent>('app-modal--edit-item--close')
    emit<AppModalAddItemCloseEvent>('app-modal--add-item--close')
    emit<AppModalSearchResultsCloseEvent>('app-modal--search-results--close')
    document.location.search = ''
    return true
  }

  private getItemFieldsToPush (data: FormEditFormData): AirtableRecord['fields'] {
    const fields: AirtableRecord['fields'] = {}
    if (data.barcode.length > 0) fields.barcode = data.barcode
    if (data.box.length > 0) fields.box = data.box
    if (data.brand.length > 0) fields.brand = data.brand
    if (data.category.length > 0) fields.category = data.category
    if (data.details.length > 0) fields.details = data.details
    if (data.drawer.length > 0) fields.drawer = data.drawer
    if (data.location.length > 0) fields.location = data.location
    if (data.name.length > 0) fields.name = data.name
    if (data.photo) fields.photo = [{ url: data.photo } as ItemPhoto] // we don't need the whole object
    if (data.price !== undefined) fields.price = data.price
    if (data.reference.length > 0) fields.reference = data.reference
    if (data.status.length > 0) fields.status = data.status
    fields['ref-printed'] = data['ref-printed']
    logger.log('fields before clean', copy(fields))
    if (data.id) {
      const existing = this.items.find(existingItem => existingItem.id === data.id)
      if (!existing) throw new Error('existing item not found locally')
      const dataFields = Object.keys(fields) as ItemField[]
      dataFields.forEach((field) => {
        if (field === ItemField.id) return
        const samePhoto = (field === ItemField.photo && existing.photo && existing.photo[0]?.url === fields.photo?.[0]?.url) ?? false
        const sameValue = existing[field] === fields[field]
        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
        if (samePhoto || sameValue) delete fields[field]
      })
      logger.log('fields after clean', copy(fields))
    }
    return fields
  }

  private async pushItemRemotely (fields: AirtableRecord['fields'], id: Item['id']): Promise<AirtableRecord> {
    const data = { fields }
    if (id === '') return post(this.apiUrl, data)
    const url = this.apiUrl.replace('?', `/${id}?`)
    return patch(url, data)
  }

  private pushItemLocally (itemTouched: Item): void {
    logger.log('pushing item locally', itemTouched)
    const index = this.items.findIndex(item => item.id === itemTouched.id)
    if (index >= 0) this.items[index] = itemTouched // update existing item
    else if (itemTouched.id) this.items.push(itemTouched) // new item with id
    else logger.showError('cannot add item without id')
    this.initFuse()
  }

  private handleActions (): void {
    document.addEventListener('click', (event) => {
      if (!(event.target instanceof HTMLElement)) return
      const { target } = event
      let action = target.dataset['action']
      let payload = target.dataset['payload'] as object | string | undefined
      if (action === undefined) return
      action = action.trim()
      logger.log('action clicked :', action)
      if (typeof payload === 'string' && payload.startsWith('{')) payload = JSON.parse(payload) as object
      if (payload !== undefined) logger.log('payload :', payload)
      event.stopPropagation()
      emit<AppActionEvent>(action, payload ?? target)
    })
  }

  private saveCommonLists (lists: CommonLists): void {
    logger.log('saving common lists :', lists)
    const names = Object.keys(lists) as (keyof CommonLists)[]
    names.forEach((name) => {
      lists[name] = ['', ...lists[name].sort((a, b) => Intl.Collator().compare(a, b))]
    })
    storage.set('lists', lists)
  }

  private readCommonLists (): boolean {
    if (this.commonListsLoaded) return true
    const lists = storage.get<CommonLists>('lists', EMPTY_COMMON_LISTS)
    logger.log('common lists found', lists)
    const template = find.one('template#edit-item')
    const data: Record<keyof CommonLists, string> = {
      boxes: valuesToOptions(lists.boxes),
      locations: valuesToOptions(lists.locations),
      statuses: valuesToOptions(lists.statuses),
      drawers: valuesToOptions(lists.drawers),
      categories: valuesToOptions(lists.categories),
    }
    template.innerHTML = fillTemplate(template.innerHTML, data)
    this.commonListsLoaded = true
    return true
  }
}

new App()
