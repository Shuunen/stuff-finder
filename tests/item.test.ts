import { expect, it } from 'vitest'
import { ItemStatus, type AirtableRecord, type Item } from '../src/types'
import { airtableRecordToItem, fakeItem, getCommonListsFromItems } from '../src/utils/item.utils'

const recordA: AirtableRecord = {
  fields: {
    box: 'box A',
    category: 'category A',
    location: 'location A',
    name: 'item A',
  },
  id: 'rec123',
}

it('airtableRecordToItem A', () => {
  expect(airtableRecordToItem(recordA)).toMatchSnapshot()
})

const itemBase: Item = {
  'barcode': 'barcode B',
  'box': 'box B',
  'brand': 'brand B',
  'category': 'category B',
  'details': 'details B',
  'drawer': 'drawer B',
  'id': 'rec234',
  'location': 'location B',
  'name': 'name B',
  'ref-printed': false,
  'reference': 'reference B',
  'status': ItemStatus.Achet√©,
  'updated-on': '2021-08-01T00:00:00.000Z',
}

function createFakeItem (data: Partial<Item>) {
  return { ...itemBase, ...data }
}

// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
const itemA = createFakeItem({ id: 'itemA', status: 'new surprise status' as ItemStatus })
const itemB = createFakeItem({ id: 'itemB', location: 'location A' })

it('getCommonListsFromItems A', () => {
  expect(getCommonListsFromItems([itemA, itemB])).toMatchSnapshot()
})

it('fakeItem A', () => {
  const name = 'super glass tempered shovel'
  const item = fakeItem(name)
  expect(item.name).toBe(name)
  expect(item.photo).toHaveLength(1)
})
