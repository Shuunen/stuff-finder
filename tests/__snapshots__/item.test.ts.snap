// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`airtableRecordToItem A 1`] = `
{
  "barcode": "",
  "box": "box A",
  "brand": "",
  "category": "category A",
  "details": "",
  "drawer": "",
  "id": "rec123",
  "location": "location A",
  "name": "item A",
  "photo": [],
  "price": 0,
  "ref-printed": false,
  "reference": "",
  "status": "acheté",
  "updated-on": "",
}
`;

exports[`getAllItems A no offset 1`] = `
{
  "records": [
    {
      "fields": {
        "barcode": "",
        "box": "box A",
        "brand": "",
        "category": "category A",
        "details": "",
        "drawer": "",
        "location": "location A",
        "name": "item A",
        "ref-printed": false,
        "reference": "",
        "status": "acheté",
        "updated-on": "",
      },
      "id": "rec123",
    },
  ],
}
`;

exports[`getAllItems B with offset 1`] = `
{
  "records": [
    {
      "fields": {
        "barcode": "",
        "box": "box A",
        "brand": "",
        "category": "category A",
        "details": "",
        "drawer": "",
        "location": "location A",
        "name": "item A",
        "ref-printed": false,
        "reference": "",
        "status": "acheté",
        "updated-on": "",
      },
      "id": "rec123",
    },
  ],
}
`;

exports[`getCommonListsFromItems A 1`] = `
{
  "boxes": [
    "",
    "box B",
  ],
  "categories": [
    "",
    "category B",
  ],
  "drawers": [
    "",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
  ],
  "locations": [
    "",
    "location A",
    "location B",
  ],
  "statuses": [
    "",
    "à donner",
    "à vendre",
    "acheté",
    "donné",
    "jeté",
    "new surprise status",
    "renvoyé",
    "vendu",
  ],
}
`;

exports[`getOneItem A 1`] = `
{
  "barcode": "",
  "box": "box A",
  "brand": "",
  "category": "category A",
  "details": "",
  "drawer": "",
  "id": "rec123",
  "location": "location A",
  "name": "item A",
  "photo": [],
  "price": 0,
  "ref-printed": false,
  "reference": "",
  "status": "acheté",
  "updated-on": "",
}
`;

exports[`pushItemRemotely A itemAA updated && no id should post 1`] = `
{
  "fields": {
    "barcode": "",
    "box": "box A",
    "brand": "",
    "category": "category A",
    "details": "",
    "drawer": "",
    "location": "location A",
    "name": "item A",
    "ref-printed": false,
    "reference": "",
    "status": "acheté",
    "updated-on": "",
  },
  "id": "rec123",
}
`;

exports[`pushItemRemotely A itemAA updated && no id should post 3`] = `
{
  "fields": {
    "status": "acheté",
  },
}
`;

exports[`pushItemRemotely B itemAA updated && his id should patch 1`] = `
{
  "fields": {
    "barcode": "",
    "box": "box A",
    "brand": "",
    "category": "category A",
    "details": "",
    "drawer": "",
    "location": "location A",
    "name": "item A",
    "ref-printed": false,
    "reference": "",
    "status": "acheté",
    "updated-on": "",
  },
  "id": "rec123",
}
`;

exports[`pushItemRemotely B itemAA updated && his id should patch 3`] = `
{
  "fields": {
    "status": "acheté",
  },
}
`;
